internal/cmd/restore.go

```diff
@@ -0,0 +1,139 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/jzelinskie/cobrautil"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/cobra"
+
+	"github.com/authzed/zed/internal/backupformat"
+	"github.com/authzed/zed/internal/client"
+)
+
+func registerRestoreCmd(rootCmd *cobra.Command) {
+	rootCmd.AddCommand(restoreCmd)
+	restoreCmd.Flags().Int("batch-size", 1_000, "restore relationship write batch size")
+	restoreCmd.Flags().Int("batches-per-transaction", 10, "number of batches per transaction")
+}
+
+var restoreCmd = &cobra.Command{
+	Use:   "restore <filename>",
+	Short: "Restore a permission system from a file",
+	Args:  cobra.ExactArgs(1),
+	RunE:  restoreCmdFunc,
+}
+
+func restoreCmdFunc(cmd *cobra.Command, args []string) error {
+	filename := args[0]
+
+	log.Trace().Str("filename", filename).Send()
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return fmt.Errorf("unable to open restore file: %w", err)
+	}
+
+	decoder, err := backupformat.NewDecoder(f)
+	if err != nil {
+		return fmt.Errorf("error creating restore file decoder: %w", err)
+	}
+
+	client, err := client.NewClient(cmd)
+	if err != nil {
+		return fmt.Errorf("unable to initialize client: %w", err)
+	}
+
+	ctx := context.Background()
+
+	if _, err := client.WriteSchema(ctx, &v1.WriteSchemaRequest{
+		Schema: decoder.Schema(),
+	}); err != nil {
+		return fmt.Errorf("unable to write schema: %w", err)
+	}
+
+	log.Debug().Msg("schema written")
+
+	relationshipWriteStart := time.Now()
+
+	relationshipWriter, err := client.BulkImportRelationships(ctx)
+	if err != nil {
+		return fmt.Errorf("error creating writer stream: %w", err)
+	}
+
+	batchSize := cobrautil.MustGetInt(cmd, "batch-size")
+	batchesPerTransaction := cobrautil.MustGetInt(cmd, "batches-per-transaction")
+
+	batch := make([]*v1.Relationship, 0, batchSize)
```

Comment:
This batching logic looks good, but I think SpiceDB has a generic batcher that might be more readable. I'm not sure if it's worth pulling in.

Reply:
Does it work when you don't have everything in a slice already to start?

Reply:
Nope -- https://github.com/authzed/spicedb/blob/2ead4675d99e9a2a41dfa6f536b5079c9c626032/pkg/util/chunking.go#L7